ccan/ccan/graphql/graphql.c

```diff
@@ -0,0 +1,1031 @@
+		if (punct == 0x2026)
+			sprintf(punctbuf, "expected: '...'");
+		else
+			sprintf(punctbuf, "expected: '%c'", punct);
+		*err = punctbuf; return NULL;
+	}
+	CONSUME_ONE;
+	return tok;
+}
+
+RET parse_name(PARAMS) {
+	struct graphql_token *tok = list_top(tokens, struct graphql_token, list);
+	if (!tok || tok->token_type != 'a') {
+		*err = "name expected"; return NULL;
+	}
+	CONSUME_ONE;
+	return tok;
+}
+
+RET parse_int(PARAMS) {
+	struct graphql_token *tok = list_top(tokens, struct graphql_token, list);
+	if (!tok || tok->token_type != 'i') {
+		*err = "integer expected"; return NULL;
+	}
+	CONSUME_ONE;
+	return tok;
+}
+
+RET parse_float(PARAMS) {
+	struct graphql_token *tok = list_top(tokens, struct graphql_token, list);
+	if (!tok || tok->token_type != 'f') {
+		*err = "float expected"; return NULL;
+	}
+	CONSUME_ONE;
+	return tok;
+}
+
+RET parse_string(PARAMS) {
+	struct graphql_token *tok = list_top(tokens, struct graphql_token, list);
+	if (!tok || tok->token_type != 's') {
+		*err = "string expected"; return NULL;
+	}
+	CONSUME_ONE;
+	return tok;
+}
+
+
+// Convert input string into tokens.
+const char *graphql_lex(const char *input, const tal_t *ctx, struct list_head **tokens) {
+
+	unsigned int c;
+	const char *p, *line_beginning;
+	unsigned int line_num = 1;
+	struct list_head *tok_list;
+	struct graphql_token *tok;
+
+	// Initialize token output list.
+	tok_list = tal(ctx, struct list_head);
+	if (tokens)
+		*tokens = tok_list;
+	list_head_init(tok_list);
+
+	// Note: label and goto are used here like a continue statement except that
+        // it skips iteration, for when characters are fetched in the loop body.
+	p = input;
+	line_beginning = p;
+	do {
+		c = *p++;
+newchar:
+		// Consume line terminators and increment line counter.
+		if (LINE_TERMINATOR(c)) {
+			unsigned int c0 = c;
+			c = *p++;
+			if (c0 == 10 || c0 == 13)
+				line_num++;
+			if (c0 == 13 && c == 10)
+				c = *p++;
+			line_beginning = p - 1;
+			goto newchar;
+		}
+
+		// Consume other ignored tokens.
+		if (COMMA(c) || WHITE_SPACE(c)) {
+			c = *p++;
+			goto newchar;
+		}
+		if (COMMENT(c)) {
+			while ((c = *p++) != EOF && !EOF_CHAR(c) && COMMENT_CHAR(c))
+				; // No-op
+			goto newchar;
+		}
+
+		// Return success when end is reached.
+		if (EOF_CHAR(c))
+			return GRAPHQL_SUCCESS;
+
+		// Punctuator tokens.
+		if (PUNCTUATOR(c)) {
+
+			// Note beginning of token in input.
+			const char *start = p - 1;
+
+			// Handle the ... multi-character case.
+			if (c == '.') {
+				c = *p++;
+				if (c != '.')
+					return "unrecognized punctuator";
+				c = *p++;
+				if (c != '.')
+					return "unrecognized punctuator";
+				c = 0x2026;
+			}
+
+			tok = tal(tok_list, struct graphql_token);
+			list_add_tail(tok_list, &tok->list);
+			tok->token_type = c;
+			tok->token_specific = c;
+			tok->token_string = NULL;
+			tok->source_line = line_num;
+			tok->source_column = start - line_beginning + 1;
+			tok->source_len = p - start;
```

Comment:
Tempting to suggest a new_tok(tok_list, ...) function?

Reply:
Good idea. (Not yet implemented.)